File: examples/fibonacci.js
import { run, bench, summary } from '..';

function fibonacci(n) {
  if (n <= 1) return n;
  let a = 0, b = 1;
  for (let i = 2; i <= n; i++) {
    [a, b] = [b, a + b];
  }
  return b;
}

function fibonacciRecursive(n) {
  if (n <= 1) return n;
  return fibonacciRecursive(n - 1) + fibonacciRecursive(n - 2);
}

function fibonacciIterative(n) {
  if (n <= 1) return n;
  let a = 0, b = 1;
  for (let i = 2; i <= n; i++) {
    [a, b] = [b, a + b];
  }
  return b;
}

summary(() => {
  bench('recursive fibonacci (20)', () => {
    fibonacciRecursive(20);
  });

  bench('iterative fibonacci (20)', () => {
    fibonacciIterative(20);
  });

  bench('fibonacci($n)', function* (state) {
    const n = state.get('n');
    yield () => fibonacci(n);
  }).range('n', 10, 40, 2);
});

await run();

File: examples/fibonacci.cpp
#include "../src/mitata.hpp"

int fibonacci(int n) {
  if (n <= 1) return n;
  return fibonacci(n - 1) + fibonacci(n - 2);
}

int main() {
  mitata::runner runner;

  runner.bench("noop", []() { })->compact(true);

  runner.bench("fibonacci", []() {
    fibonacci(20);
  });

  runner.group([&]() {
    runner.bench("noop2", []() { })->compact(false);
    runner.bench("fibonacci2", []() { fibonacci(20); });
  });

  runner.summary([&]() {
    runner.bench("noop3", []() { })->compact(false);
    runner.bench("fibonacci3", []() { fibonacci(20); })->baseline(false);
    runner.bench("fibonacci4", []() { fibonacci(20); })->baseline(true);
  });

  runner.bench("noop4", []() { })->compact(true);

  auto stats = runner.run({ .colors = true, .format = "mitata", .filter = std::regex(".*") });
}